### PR TITLE
Better tagging and management of adhoc environments

### PR DESCRIPTION
**Pull request by cat5inthecradle**
In an effort to better manage adhocs and reduce costs by terminating old adhoc environments.

This PR replaces the "stop_inactive_adhoc_instances" with a more elaborate "cleanup_adhoc_evironments" cron task. This task performs three actions:

* stop_running_adhoc_instances: all adhoc ec2 instances that don't have termination protection enabled will be stopped and tagged
* delete_stale_adhoc_stacks: all adhoc ec2 instances tagged in the previous step more than 6 days ago will have their stacks deleted
* log_error_stacks: finally, any cloudformation stacks in an error state will be sent to the infra team. Note that the previous step does not wait for success, so failures may not be reported until next week when this is run again.

## Testing story

before merging, I'd like to run this manually to confirm it's behavior.

## Deployment strategy

merge and monitor the first automated execution.

### CODE REVIEW

File: bin/cron/cleanup_adhoc_environments [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    stack_name = instance.tags.find {|tag| tag.key == 'aws:cloudformation:stack-name'}&.value
+    next if stack_name.nil?
+
+    stacks_to_delete << stack_name
+  end
+
+  if stacks_to_delete.any? && !$dry_run
+    ChatClient.message 'adhoc', "Deleting CloudFormation Stacks:\n" + stacks_to_delete.join("\n")
+    stacks_to_delete.each {|stack_name| cloudformation.delete_stack(stack_name: stack_name)}
+  end
```
Comment (cat5inthecradle) [inline bin/cron/cleanup_adhoc_environments]: deleting cloudformation stacks is scary. This could absolutely attempt to delete critical stacks if their names ended up in the list. Interested in any suggestions for additional security.

We could perhaps put an exclusion in the AWS IAM Role being used here to explicitly disallow stack deletion for several critical stacks.

File: bin/cron/cleanup_adhoc_environments [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+require 'aws-sdk-ec2'
+require 'aws-sdk-cloudformation'
+require 'cdo/chat_client'
+
+$dry_run = !rack_env?(:production)
+
+SECONDS_IN_A_DAY = 24 * 60 * 60
+
+def stop_running_adhoc_instances
+  adhocs = get_adhoc_instances('running')
```
Comment (pablo-code-org) [inline bin/cron/cleanup_adhoc_environments]: always a good practice if we had this as a const ('running')

or as a global enum

File: bin/cron/cleanup_adhoc_environments [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  adhocs.each do |instance|
+    # Check if instance was stopped more than 6 days ago
+    stopped_at_tag = instance.tags.find {|tag| tag.key == 'adhoc_stopped_at'}
+    next if stopped_at_tag.nil?
+    stopped_at = DateTime.parse(stopped_at_tag.value)
+
+    next unless more_than_six_days_ago?(stopped_at)
+
+    stack_name = instance.tags.find {|tag| tag.key == 'aws:cloudformation:stack-name'}&.value
+    next if stack_name.nil?
```
Comment (pablo-code-org) [inline bin/cron/cleanup_adhoc_environments]: I would wrap this into a method called `delete_stale_adhoc_stack`

The individual step allows a developer to just analyze this and also to reuse the code if needed.

Comment (pablo-code-org) [reply bin/cron/cleanup_adhoc_environments]: Can we have a class

AdhocInstance

def stop
end

File: bin/cron/cleanup_adhoc_environments [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  end
+end
+
+# Returns true if the given time is more than six days ago
+def more_than_six_days_ago?(time)
+  (Time.now - time) > 6 * SECONDS_IN_A_DAY
+end
+
+# Returns a collection of adhoc instances
+def get_adhoc_instances(state = nil)
```
Comment (pablo-code-org) [inline bin/cron/cleanup_adhoc_environments]: This should be part of a bigger class (AdhocInstance)
This will allow code de-duplication. If you attach it to a script, it ill be only useful here.

File: bin/cron/cleanup_adhoc_environments (hunk 2 above)
Comment (pablo-code-org) [reply bin/cron/cleanup_adhoc_environments]: Also this could be part of a class/library where you pass the "state".  I wouldn't print anything inside that method. 

I would make the method return any instances and then print the message

```
 msg = "Stopping #{adhocs.length} adhoc instances:\n" +
    adhocs.map {|a| a.tags.detect {|t| t.key == 'Name'}&.value}.join("\n")
```

File: lib/cdo/rake_utils.rb [hunk 5]
```diff
@@ -169,6 +169,14 @@ def self.git_add(*args)
...
   end
 
+  def self.git_user_name
+    `git config --get user.name`
+  end
+
+  def self.git_user_email
+    `git config --get user.email`
+  end
+
```
Comment (pablo-code-org) [inline lib/cdo/rake_utils.rb]: This belongs to a different PR :P 

Making the base commit that other one will remove these files on the changes but it wouldn't run done :(

**Review by pablo-code-org — COMMENTED** [review]
I wanna `request changes` but I'll add this as `comment`

File: bin/cron/cleanup_adhoc_environments (hunk 1 above)
Comment (Hamms) [reply bin/cron/cleanup_adhoc_environments]: Is there logic we could add here that does something like verify every EC2 instance in the stack is already in the stopped state? Or an "adhoc" tag that we apply to stack that we could check for?

I feel like we have a high level of confidence in this code that the _instances_ we're interacting with are all adhoc instances, but less confidence that the stack here is necessarily an adhoc stack.

File: bin/cron/cleanup_adhoc_environments (hunk 2 above)
Comment (cat5inthecradle) [reply bin/cron/cleanup_adhoc_environments]: did some refactoring to stop passing strings around.

File: bin/cron/cleanup_adhoc_environments (hunk 4 above)
Comment (cat5inthecradle) [reply bin/cron/cleanup_adhoc_environments]: I see your point... gonna ponder if I want to go to that effort. I'm not sure where that file should live.

File: bin/cron/cleanup_adhoc_environments (hunk 1 above)
Comment (cat5inthecradle) [reply bin/cron/cleanup_adhoc_environments]: I added a check to ensure we're not deleting any of our most critical stacks. That should never raise because we are selecting ec2 instances with the environment tag 'adhoc' and then discovering what stack they belong to.

File: bin/cron/cleanup_adhoc_environments (hunk 4 above)
Comment (cat5inthecradle) [reply bin/cron/cleanup_adhoc_environments]: kept this 'script' style, vs class style.

**Comment by sureshc** [comment]
Instead of iterating over EC2 Instances that have the tag `environment` with the value `adhoc`, should we iterate over Stacks (with `environment == adhoc`)? By default, adhoc deployments have a single EC2 Instance, but a full adhoc deployment could have a console, daemon, AMIBuilder, and multiple FrontEnd EC2 Instances. Do we want to stop all the EC2 Instances in an adhoc Stack? Or is it sufficient to just stop the Web Application servers? If the daemon is still running, a new commit pushed to the feature branch might rebuild the Stack and resurrect stopped web application server instances?

**Comment by sureshc** [comment]
I'd love to take this moment to update the `adhoc:start` task to track which authenticated AWS user/role created the Stack. That might help with notifications as adhoc Stacks are scheduled for being stopped/deleted.

**Comment by cat5inthecradle** [comment]
> I'd love to take this moment to update the `adhoc:start` task to track which authenticated AWS user/role created the Stack. That might help with notifications as adhoc Stacks are scheduled for being stopped/deleted.

We already do, but I'm suggesting changing that in this PR. See my changes to cdo_app.rb

**Comment by cat5inthecradle** [comment]
> Instead of iterating over EC2 Instances that have the tag `environment` with the value `adhoc`, should we iterate over Stacks (with `environment == adhoc`)?

This PR preserves the current behavior for detecting adhoc ec2 instances. I think it also behaves like you describe, as it's also requiring the instance to have `'tag:aws:cloudformation:logical-id' => 'WebServer'`. Does the current cron task stop daemon/console/etc instances?

**Comment by sureshc** [comment]
> > I'd love to take this moment to update the `adhoc:start` task to track which authenticated AWS user/role created the Stack. That might help with notifications as adhoc Stacks are scheduled for being stopped/deleted.
> 
> We already do, but I'm suggesting changing that in this PR. See my changes to cdo_app.rb

oops. I missed that. Apologies for not reading closely!

File: lib/cdo/cloud_formation/cdo_app.rb [hunk 6]
```diff
@@ -75,7 +75,7 @@ def initialize(**options)
 
       log_resource_filter.push 'FrontendLaunchConfig', 'ASGCount'
       tags.push(key: 'environment', value: rack_env)
-      tags.push(key: 'owner', value: Aws::STS::Client.new.get_caller_identity.arn) if rack_env?(:adhoc)
+      tags.push(key: 'owner', value: Aws::STS::Client.new.get_caller_identity.arn.split(':').last) if rack_env?(:adhoc)
```
Comment (sureshc) [inline lib/cdo/cloud_formation/cdo_app.rb]: It’s may be more verbose, but we could use the ARN Parser:

`Aws::ARNParser.parse(Aws::STS::Client.new.get_caller_identity.arn).resource`

Comment (cat5inthecradle) [reply lib/cdo/cloud_formation/cdo_app.rb]: I like that better

Comment (cat5inthecradle) [reply lib/cdo/cloud_formation/cdo_app.rb]:
```suggestion
      tags.push(key: 'owner', value: Aws::ARNParser.parse(Aws::STS::Client.new.get_caller_identity.arn).resource) if rack_env?(:adhoc)
```

**Comment by sureshc** [comment]
> This PR preserves the current behavior for detecting adhoc ec2 instances. I think it also behaves like you describe, as it's also requiring the instance to have `'tag:aws:cloudformation:logical-id' => 'WebServer'`. Does the current cron task stop daemon/console/etc instances?

This scheduled process [originally iterated over adhoc Stacks](https://github.com/code-dot-org/code-dot-org/blob/8cd521791655c148002acbcd31c31606c278c33e/bin/cron/stop_inactive_adhoc_instances#L49-L55), and I’m suggesting we revert back to that logic. Our adhoc provisioning and management tasks too often assume that the adhoc environment is a single EC2 Instance, and I’m suggesting we intentionally recognize that an adhoc is a Stack with many Resources, with possibly multiple EC2 Instances if the full architecture is configured: a console server, a daemon server, an AMI Builder, and multiple load-balanced WebServer instances. The current job does not stop daemon/console servers if a full architecture is configured for an adhoc.

**Comment by cat5inthecradle** [comment]
Making this stack-focused makes perfect sense. Thanks for that nudge. I think we should conceptually approach this as:

1. 'spin down' the adhoc stack every friday night, as a cost-savings measure. Must be easily spin-up-able
2. destroy adhoc stacks that have been 'spun-down' for a week or more.

The way we determine 'can be destroyed' is by answering the question 'did anyone notice that we spun it down'.

**Comment by sureshc** [comment]
Suggestion for follow up work: stop the Aurora cluster in the Stack to constrain costs. I think RDS costs are db instance minutes + storage + IOPS? storage and IOPS are going to be low for an adhoc. Looks like [we default](https://github.com/code-dot-org/code-dot-org/blob/a13c6d622726dbc8795961d8b4b07ed5e6b3b20b/aws/cloudformation/cloud_formation_stack.yml.erb#L30) to `db.r5.2xlarge` for any environment that has a database. Defaulting to `db.t3.micro` (or `small`) for adhocs might mitigate our adhoc RDS costs. We could also disable automated Snapshots for adhocs. Again ... this is a follow up suggestion. Just jotting it down here so I don't forget.
